### ISSUE TITLE
Fix `no_cache` feature and recompute fees

### ISSUE THREAD

**Issue by nearmax** [near/nearcore#2948]
According to https://github.com/nearprotocol/nearcore/issues/2778#issuecomment-638506369 `no_cache` feature sets the cache size to 1 instead of disabling it entirely. This most likely leads to incorrect computation of fees, especially the newly added fee for loading/compiling the contract. We need to fix `no_cache` feature to disable the cache and recompute the fees

**Comment by bowenwang1996** [comment]
Is `no_cache` on by default?

**Comment by willemneal** [comment]
@bowenwang1996 Yep
```toml
near-vm-runner = {path = "../../runtime/near-vm-runner" , features = ["costs_counting", "no_cache"]}
node-runtime = { path = "../../runtime/runtime" , features = ["costs_counting", "no_cache"]}
```

**Comment by willemneal** [comment]
According to the vm-runner it doesn't use the cache.

```rust
#[cfg(feature = "no_cache")]
pub(crate) fn compile_module(
    _code_hash: Vec<u8>,
    code: &[u8],
    config: &VMConfig,
) -> Result<wasmer_runtime::Module, VMError> {
    let prepared_code = prepare::prepare_contract(code, config)?;
    wasmer_runtime::compile(&prepared_code).map_err(|err| err.into_vm_error())
}

```

Regardless the values of how much it costs to load and compile are done by calling each VMKind's compile method directly.  And upon inspection the new runner for wasmtime doesn't use the cache at all.

It seems the cache size of 1 refers to the trie, which would only affect the performance if the same entry was read twice in a row.  And this value cannot be zero.

**Comment by nearmax** [comment]
> Is `no_cache` on by default?

It is not by default for the regular node.

**Comment by nearmax** [comment]
CC @olonho

**Comment by bowenwang1996** [comment]
cc @matklad

**Comment by matklad** [comment]
To clarify, `no_cache` affects to orthogonal caches:

* `ChainStore` [cache](https://github.com/near/nearcore/blob/f4c194f700f856be823a979b6870350762366feb/chain/chain/src/store.rs#L54-L64) which caches access to rocksdb storeage
* [In-memory contract compilation cache](https://github.com/near/nearcore/blob/f4c194f700f856be823a979b6870350762366feb/runtime/near-vm-runner/src/cache.rs#L187-L202), which stores compiled and deserialized contract.

For the second cache, the behavior of `no_cache` feature is correct, and it completely disables the caching. The `ChainStore` cache indeed retains capacity of 1.

I don't think that `ChainStore` cache affects the [compilation cost fee](https://github.com/near/nearcore/blob/f4c194f700f856be823a979b6870350762366feb/runtime/runtime-params-estimator/src/vm_estimator.rs#L202) specifically, as it now doesn't do repeated fetches of the same key. 

Not sure how `ChainStore` cache affects the overall cost. I think it's worth to invest time into figuring out the way to remove the cache completely, and see how it affects the costs.

**Comment by bowenwang1996** [comment]
It also affects `TrieCache`

**Comment by bowenwang1996** [comment]
> Not sure how ChainStore cache affects the overall cost

@matklad is `ChainStore` even used at all for the cost estimation?

**Comment by matklad** [comment]
cc @jakmeier i am curious if there's some action items here, or shoudld we just close? It *might* be that cache of size one is worst-case enough? 

Curiously, my recent work on non-`&mut` caches probably allows us to actually set the cache size to zero, as we no longer return refs into the cache, so, at minimum, we should try to set capatity to zero and run estimator.

**Comment by jakmeier** [comment]
@matklad I actually disagree with the premise of this issue.

1. I start to believe we should be running with some caches enabled. Much of our code relies on it and will make no effort to call for example `get_account` as few times as possible, as the subsequent calls are assumed to be served from memory. Forcing those to go to disk every time is not in the spirit of worst-case we are aiming for.
2. The aforementioned contract cache issue is better solved by not reusing the same contract code, therefore making caching impossible. This particular issue has been fixed in #6208. Generally speaking, I would rather enforce cache misses on a case-by-case basis in each estimation.
3. True worst-case for a cached system is not having no cache. Usually it is something like accessing a cache, missing, evicting an older value, and inserting the new value. Granted, compared to IO, evicting non-dirty data is irrelevant. So I wouldn't worry too much for any of the estimations we have today. But again, this is something we have to consider for each estimation separately.

So yeah if you ask me, we can close this.
